.github/actions/upload/action.yml

```diff
@@ -1,31 +1,41 @@
 name: upload
-description: "Upload test results"
+description: "Upload test results, reports, and heap dump."
 inputs:
   name-suffix:
     default: ${{ github.job }}
+  upload-heap-dump:
+    default: false
   test-report-retention-days:
     default: 5
+  heap-dump-retention-days:
+    default: 14
 
 runs:
   using: composite
   steps:
     - name: Upload test results
       uses: actions/upload-artifact@v3
-      # Upload all test reports only on failure, because the artifacts are large
-      if: failure()
+      if: inputs.upload-results
       with:
         name: result ${{ inputs.name-suffix }}
+        if-no-files-found: 'ignore'
```

Comment:
this is not same as uploading only on failure - the files get written if tests run AFAIK.

Reply:
Oh right. I can't use the `failure()` in any expression for input parameters, only in the `if:` block. I'd have to assign IDs to every step and use `step.some-id.conclusion == 'failure'`, there's one job where we have many steps and this will be a long expression :-(